### PR TITLE
Eliminate Effect::mDurationFormat; New event handling for DTMF generator

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: #2673
Resolves: #2674

To eliminate EffectHostInterface and state variables in Effect that it accessed and modified has involved some unusual twists and turns...  The dependencies among the necessary tasks are becoming surprising on closer look.  This big PR gives
necessary solutions to various sub-sub-tasks.

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/effects/Effect.cpp [hunk 1]
```diff
@@ -1327,7 +1321,7 @@ bool Effect::ProcessTrack(EffectSettings &settings,
    decltype(mBufferSize) outputBufferCnt = 0;
    bool cleared = false;
 
-   auto chans = std::min<unsigned>(mNumAudioOut, mNumChannels);
+   auto chans = std::min<unsigned>(GetAudioOutCount(), mNumChannels);
```
Comment (pietro68) [inline src/effects/Effect.cpp:1288]: Could `numAudioOut` be used instead?

File: libraries/lib-components/EffectInterface.h [hunk 2]
```diff
@@ -64,9 +64,20 @@ typedef enum EffectType : int
 
 using EffectFamilySymbol = ComponentInterfaceSymbol;
 
+//! Non-polymorphic package of settings values common to many effects
+class EffectSettingsExtra final {
+};
```
Comment (pietro68) [inline libraries/lib-components/EffectInterface.h:76]: Is this a placeholder that will be filled later on?

Comment (pietro68) [reply libraries/lib-components/EffectInterface.h:76]: The next commit answered my question.

File: src/effects/DtmfGen.cpp [hunk 3]
```diff
@@ -568,33 +578,41 @@ bool EffectDtmf::MakeDtmfTone(float *buffer, size_t len, float fs, wxChar tone,
...
+   settings.Recalculate(*this);
    UpdateUI();
 }
 
 void EffectDtmf::OnDuration(wxCommandEvent & WXUNUSED(evt))
 {
+   auto &settings = mSettings;
    SetDuration(mDtmfDurationT->GetValue());
-   Recalculate();
+   settings.Recalculate(*this);
```
Comment (pietro68) [inline src/effects/DtmfGen.cpp:608]: Can we not call this on mSettings directly?

**Review by pietro68 — APPROVED** [review]
I looked mainly at the last 6 commits (as told to) and I think I understand what goes on there. I only added some minor observations.

File: src/effects/DtmfGen.cpp (hunk 3 above)
Comment (Paul-Licameli) [reply src/effects/DtmfGen.cpp:608]: I anticipate that the means of communicating the settings into many of these functions might change might change in the future.  Binding a reference to the settings at the top will avoid the need to rewrite many lines again later.  I did that consistently, even in some short functions.

**Comment by Paul-Licameli** [comment]
Pietro, please review it again.

It is rebased onto recent master.  The most recent four commits are new.  In the one before those "Define and use Settings for EffectDtmf, but not yet externalized..." I only made variable name changes.  I use `dtmfSettings` instead of `settings` in many places.  I may want the name `settings` in the same scope later for something else.

File: src/effects/EffectUI.cpp [hunk 4]
```diff
@@ -207,7 +207,12 @@ EffectUIHost::~EffectUIHost()
 bool EffectUIHost::TransferDataToWindow()
 {
    // Transfer-to takes const reference to settings
-   return mEffectUIHost.TransferDataToWindow(mpAccess->Get());
+   return mEffectUIHost.TransferDataToWindow(mpAccess->Get()) &&
+      //! Do other apperance updates
+      mpValidator->UpdateUI() &&
+      //! Do valiators
```
Comment (pietro68) [inline src/effects/EffectUI.cpp:213]: typo (?)

**Review by pietro68 — APPROVED** [review]
All understood now!

File: src/effects/EffectUI.cpp (hunk 4 above)
Comment (Paul-Licameli) [reply src/effects/EffectUI.cpp]: thanks

**Pull request merged**